### ISSUE TITLE
SDK2 does not provide Balance query

### ISSUE THREAD

**Issue by vcyk** [pi-apps/pi-platform-docs#42]
SDK1 implemented a Balance query api for app developer to know the current balance of his/her account.  But SDK2 does not provide this.  Is this being implemented, should be easy and quick to do?  @aurelienshz

**Comment by dannybutterfield1880** [comment]
I would also like to know the answer to this question

**Comment by Sten435** [comment]
I would also like to see the balance of a user.

**Comment by B-Derouet** [comment]
For the moment, you can use the Pi Network Testnet API to get the balance of an account:

```js
const response = await fetch('https://api.testnet.minepi.com/accounts/{publicKey}')
const accountInfo = await response.json()
const balance = accountInfo.balances[0].balance
```

But be patient, I think the Core Team is actively working on the project to allow more things to the developers during the hackathon. Updates will come soon.

**Comment by dannybutterfield1880** [comment]
@B-Derouet that's great thank you for that

**Comment by Sten435** [comment]
@B-Derouet where can we find the public key?

**Comment by B-Derouet** [comment]
@Sten435 before the payment, you have to directly ask the user for his public key, but after the transaction, you can use the txid to retrieve the public key of the user:

```js
const response = await fetch('https://api.testnet.minepi.com/transactions/{txid}')
const transactionInfo = await response.json()
const userPublicKey = transactionInfo.source_account```

**Comment by dannybutterfield1880** [comment]
@B-Derouet is this the access_token that the authenticate() returns

**Comment by B-Derouet** [comment]
@dannybutterfield1880 No, this information is unrelated. As a user, you can find your public key in the wallet.pi app.

**Comment by dannybutterfield1880** [comment]
@B-Derouet Ahh ok the user's wallet address. Thank you that makes sense.

**Comment by Sten435** [comment]
@B-Derouet  where do you get all this information? Could you share the source?

**Comment by dannybutterfield1880** [comment]
@Sten435 the public key @B-Derouet was talking about is the key you get if you go to wallet.pi and press receive

**Comment by B-Derouet** [comment]
@Sten435 since Pi Network uses SCP (Stella Consensus Protocol), you can have all the information you want in the Stellar API documentation: https://developers.stellar.org/api/

The endpoint of the Pi Network Testnet API is here: https://api.testnet.minepi.com/

**Comment by browolf** [comment]
the only downside of pi is you can't send anything through the api because no one has a private key to a sign operations with.

**Comment by B-Derouet** [comment]
@browolf your private key is your wallet's mnemonic passphrase (if you use your own wallet)

**Comment by browolf** [comment]
you can't sign a transaction with a passphrase though. it has to be decrypted to get the private key

**Comment by B-Derouet** [comment]
@browolf you can get the private key from the passphrase:
```ts
import StellarSdk from 'stellar-sdk';
import bip39 from 'bip39';
import { derivePath } from '@hawkingnetwork/ed25519-hd-key-rn';

async function getPrivateKeyFromPassphrase(passphrase: string) {
  const seed = await bip39.mnemonicToSeed(passphrase);
  const derivedSeed = derivePath("m/44'/314159'/0'", seed); //derivation path of Pi Network mnemonic passphrase

  return StellarSdk.Keypair.fromRawEd25519Seed(derivedSeed.key);
}
```

**Comment by browolf** [comment]
I've tried doing it with https://iancoleman.io/bip39/ but I can't get it to give the right answer. Someone I know says it's salted

**Comment by B-Derouet** [comment]
@browolf I made this function by digging in the code of the Pi Network application and I can assure you that it works perfectly! It's normal that it doesn't return the same private key format that was given before.

**Comment by Sten435** [comment]
If anyone has questions about the Pi Network SDK or other general coding questions regarding anything related to the Pi Network, then you ask them in the Pi Network Developer Portal: https://discord.gg/qM3fXPq4fK
It is a faster way to connect with people, instead of contact people through email, or through the Pi chat app, which can be a real pain.

This is not an AD, this is an informative post for people who may need a communication server for a question.

**Comment by aurelienshz** [comment]
Hi there. Thanks @B-Derouet for providing an answer to this.

I'm sensing a small mix between two different things here:
- the `Balance` method of the SDKv1 was returning the **app's** balance - which is now simply the wallet balance of the app developer (and it can be obtained using the Testnet Blockchain API)
- the user's balance can be obtained as long as you know the user's public key.

We have decided to get rid of the `Balance` method because we wanted to keep the SDK light and concise, and it did not feel like a very relevant use case for a developer to read their wallet's balance from their app's frontend.

The end-user's balance, on the other hand, is indeed available to anyone who knows the user's public key (the blockchain is public). This might end up being exposed (with most likely an extra layer of privacy, e.g a dedicated token scope to be authorized by the user in the auth consent prompt).

There seems to be demand for higher-level abstractions around the blockchain API that aren't simply related with the user paying the app, but also the other way around (#44) - as well as data fetching, e.g account balance. I have recorded that in our internal issue tracker.
